File: numerical/windowfunc.js
/*
  Window functions. See https://en.wikipedia.org/wiki/Window_function and https://en.wikipedia.org/wiki/Mollifier
*/

exports.getNamed = getNamed;

function nuttallWindow(t) {
  if (t < 0 || t > 1) return 0;
  let tp = Math.PI*t;
  return 0.355768 - 0.487396 * Math.cos(2*tp) + 0.144232 * Math.cos(4*tp) - 0.012604 * Math.cos(6*tp);
}

function tukeyWindow(t, a) {
  if (t < 0 || t > 1) return 0;
  if (t < a/2) {
    return 0.5 * (1 + Math.cos(Math.PI * (2 * t / a - 1)));
  }
  if (t > (1-a/2)) {
    return 0.5 * (1 + Math.cos(Math.PI * (2 * (1-t) / a - 1)));
  }
  return 1;
}

function lanczosWindow(t) {
  if (t < 0 || t > 1) return 0;
  let tp = 2 * Math.PI * (t - 0.5);
  if (tp < 1e-6 && tp > -1e-6) return 1;
  return Math.sin(tp)/tp;
}

function rectangularWindow(t) {
  if (t < 0 || t > 1) return 0;
  return 1;
}

function getNamed(type) {
  if (type === 'nuttall') return nuttallWindow;
  if (type === 'tukey0.5') return (t) => tukeyWindow(t, 0.5);
  if (type === 'lanczos') return lanczosWindow;
  if (type === 'rectangular') return rectangularWindow;
  return undefined;
}

File: numerical/windowfunc.cc
#include "../common/std_headers.h"
#include "./windowfunc.h"

double nuttallWindow(double t)
{
  if (t < 0.0 || t > 1.0) return 0;
  auto tp = M_PI*t;
  return 0.355768 - 0.487396 * cos(2.0 * tp) + 0.144232 * cos(4.0 * tp) - 0.012604 * cos(6.0 * tp);
}

double tukeyWindow(double t, double a)
{
  if (t < 0.0 || t > 1.0) return 0.0;
  if (t < 0.5*a) {
    return 0.5 * (1 + cos(M_PI * (2.0 * t / a - 1.0)));
  }
  if (t > (1.0 - 0.5*a)) {
    return 0.5 * (1 + cos(M_PI * (2.0 * (1.0 - t) / a - 1.0)));
  }
  return 1.0;
}

double lanczosWindow(double t)
{
  if (t < 0.0 || t > 1.0) return 0.0;
  double tp = 2.0 * M_PI * (t - 0.5);
  if (tp < 1.0e-6 && tp > -1.0e-6) return 1.0;
  return sin(tp)/tp;
}

double rectangularWindow(double t)
{
  if (t < 0.0 || t > 1.0) return 0.0;
  return 1.0;
}

std::function< double(double t) > getNamedWindow(string const &type)
{
  if (type == "nuttall") return nuttallWindow;
  if (type == "tukey0.5") return [](double t) { return tukeyWindow(t, 0.5); };
  if (type == "lanczos") return lanczosWindow;
  if (type == "rectangular") return rectangularWindow;

  throw runtime_error("No such window");
}
